vmaas/reposcan/katello.py

```diff
@@ -0,0 +1,181 @@
+                    break
+            except requests.exceptions.RequestException:
+                tries += 1
+                LOGGER.exception("Error calling API %s: ", url)
+        self.success = False
+        return {}
+
+    def _download_katello_ca_certificate(self) -> str | None:
+        LOGGER.info("Downloading Katello CA certificate.")
+
+        ca_certificate = self._fetch_katello(KATELLO_CA_CERT_PATH, scheme="http", json=False)
+        if not ca_certificate:
+            LOGGER.error("Katello CA certificate download failed.")
+            self.success = False
+            return None
+
+        # Write CA certificate to tempfile to be able to use with requests
+        with open(os.path.join(self.tmp_directory, os.path.basename(KATELLO_CA_CERT_PATH)), "w", encoding="utf-8") as ca_cert_tmpfile:
+            ca_cert_tmpfile.write(ca_certificate["data"])
+
+        return ca_certificate["data"]
+
+    def _download_katello_access_certificate(self, org_id: int, org_label: str) -> str | None:
+        LOGGER.info("Downloading Katello organization '%s' access certificate.", org_label)
+
+        access_certificate_key = self._fetch_katello(KATELLO_ACCESS_CERT_PATH % org_id, json=False)
+        if not access_certificate_key:
+            LOGGER.error("Katello access certificate download failed, skipping organization '%s'.", org_label)
+            self.success = False
+            return None
+
+        return access_certificate_key["data"]
+
+    def _get_orgs(self) -> dict[int, str]:
+        org_results = self._fetch_katello(KATELLO_ORG_LIST_API)
+        return {result["id"]: result["label"] for result in org_results.get("results", [])}
+
+    def _get_org_repos(self, org_id: int, org_label: str, ca_cert: str, products: dict) -> list[tuple]:
+        LOGGER.info("Getting organization '%s' repositories.", org_label)
+
+        repos = []
+
+        access_cert = self._download_katello_access_certificate(org_id, org_label)
+        if not access_cert:
+            return repos
+
+        repo_results = self._fetch_katello(KATELLO_REPOS_LIST_API % org_id)
+
+        for repo in repo_results.get("results", []):
+            content_label = repo["content_label"]
+            if repo["content_type"] != "yum":
+                LOGGER.debug("Skipping repo '%s' - not a yum repo.", content_label)
+                continue
+            product_name = repo["product"]["name"]
+            cs_label = repo["content_label"]
+            cs_name = repo["name"]
+            url = repo["full_path"]
+            releasever = repo["minor"]
+
+            basearch = self._find_arch(url)
+            if not basearch:
+                LOGGER.warning("Skipping repo '%s' - unknown basearch.", content_label)
+                continue
+
+            # Update product dict with unique product names and content sets
+            if product_name not in products:
+                products[product_name] = {"product_id": None, "content_sets": {}}
```

Comment:
is product_id needed when it is always None?

Reply:
Yes, the code importing products to DB expects the key to be there